### ISSUE TITLE
trainer中没有实现eval和compute_metric

### ISSUE THREAD

**Issue by toddlt** [THUDM/ChatGLM3#376]
### System Info / 系統信息

系统环境与官方建议配置一致，本问题与环境不相关

### Who can help? / 谁可以帮助到您？

@Btlmd

### Information / 问题信息

- [ ] The official example scripts / 官方的示例脚本
- [X] My own modified scripts / 我自己修改的脚本和任务

### Reproduction / 复现过程

我把trainer改为了：

    trainer = PrefixTrainer(
        model=model,
        args=training_args,
        train_dataset=train_dataset if training_args.do_train else None,
        eval_dataset=eval_dataset if training_args.do_eval else None,
        tokenizer=tokenizer,
        data_collator=data_collator,
        compute_metrics=evaluator.compute_metrics if training_args.predict_with_generate else None,
        save_changed=model_args.pre_seq_len is not None
    )

以便在compute_metric里面计算function calling的准确和召回等指标，目前遇到点问题

### Expected behavior / 期待表现

compute_metric函数前几行这样：
    
def compute_metrics(self, eval_preds):
        preds, labels = eval_preds
        if isinstance(preds, tuple):
            preds = preds[0]

我这里拿到的preds的tensor的shape是[batch_size, max_length, vocabulary_size]，我预期的是拿到[batch_size, max_length]的，然后batch_decode之后去计算指标

请问这里我应该手动加一条torch.argmax(preds, axis=-1)来处理吗？

**Comment by Btlmd** [comment]
> ### System Info / 系統信息
> 系统环境与官方建议配置一致，本问题与环境不相关
> 
> ### Who can help? / 谁可以帮助到您？
> @Btlmd
> 
> ### Information / 问题信息
> * [ ]  The official example scripts / 官方的示例脚本
> * [x]  My own modified scripts / 我自己修改的脚本和任务
> 
> ### Reproduction / 复现过程
> 我把trainer改为了：
> 
> ```
> trainer = PrefixTrainer(
>     model=model,
>     args=training_args,
>     train_dataset=train_dataset if training_args.do_train else None,
>     eval_dataset=eval_dataset if training_args.do_eval else None,
>     tokenizer=tokenizer,
>     data_collator=data_collator,
>     compute_metrics=evaluator.compute_metrics if training_args.predict_with_generate else None,
>     save_changed=model_args.pre_seq_len is not None
> )
> ```
> 
> 以便在compute_metric里面计算function calling的准确和召回等指标，目前遇到点问题
> 
> ### Expected behavior / 期待表现
> compute_metric函数前几行这样：
> 
> def compute_metrics(self, eval_preds): preds, labels = eval_preds if isinstance(preds, tuple): preds = preds[0]
> 
> 我这里拿到的preds的tensor的shape是[batch_size, max_length, vocabulary_size]，我预期的是拿到[batch_size, max_length]的，然后batch_decode之后去计算指标
> 
> 请问这里我应该手动加一条torch.argmax(preds, axis=-1)来处理吗？

可以手动 argmax 一次看看效果，如果token符合预期的话应该是可以的

**Comment by toddlt** [comment]
我这样argmax了之后，拿到了sequence，用<|user|>和<|observation|>这俩token把它分隔成一段一段的对话，我又遇到另一个问题，要怎么判断哪里是有效对话的结束呢？

以下内容是输出文本 (里面有符合被markdown给渲染了)

> <|observation|>:
> 
> search_hotels
>  ```python
> tool_call(price_range_lower=700, price_range_upper=800)
> ```
> 
> <|user|>:
> 
>  {"name": "酒店红都维致酒店", "price": "五星型", "price": "北京",城区"直街2门大街18号", "priceway": "地铁质地站",出口出口", "price": "010-63681868", "priceilities": ["停车场设施免费私人客房"]免费ifi", "健身房网络", "停车场电话电话"] "停车场风机", "保险
> 4小时热水"] "早餐餐早餐", "咖啡餐厅", "健身房设施", "停车场游泳池"] "健身房", "早餐", "桑A", "桑烟客房", "商务中心", "早餐", "咖啡拿", "台",", "早餐",宾", "旅游",", "擦服务存", "早餐车服务 "早餐车服务", "rating": 750.00 "rating": 8.7] "ratingel_id": 100<|assistant|>
>  唐拉雅秀酒店是一家，
> 
> <|user|>:
> 
>  酒店最低在呢
> 
> <|user|>:
> 
>  这个唐拉雅秀酒店的价格是787元。
> 
> <|user|>:
> 
>  好的预订的酒店。行办理？？
> 
> <|user|>:
> 
>  您感谢，只需要拨打拨打酒店的预订010-58576688预订或者访问网站官网官网进行预订预订。
> 
> <|user|>:
> 
> 
> 
> <|user|>:
> 
> 更多关于更多关于更多关于更多关于更多关于更多关于
> 
> <|user|>:
> 
> 更多关于更多关于更多关于更多关于更多关于更多关于更多关于更多关于更多关于更多关于更多关于更多关于 是非常非常重要的 是非常非常重
> 要的
> 
> <|user|>:
> 
> <|assistant|>更多关于更多关于 是非常非常重要的 是非常非常重要的 是非常非常重要的 是非常非常重要的 是非常非常重要的 是非常非常重
> 要的 是非常非常重要的 是非常非常重要的 是非常非常重要的 是非常非常重要的 是非常非常重要的 是非常非常重要的 是非常非常重要的更多
> 关于更多关于更多关于更多关于 是非常非常重要的 是非常非常重要的 是非常非常重要的 是非常非常重要的更多关于
> 
> <|user|>:
> 
> 是非常非常重要的 是非常非常重要的 是非常非常重要的 是非常非常重要的更多关于更多关于更多关于更多关于更多关于 ``` ``` ``` ``` ``` ``` ```更多关于更多关于更多关于 是非常非常重要的 ``` ``` ``` ``` ``` ``` ``` ``` ``` ``` ``` ```更多关于更多关于 ``` ``` ``` ``` ``` ``` ``` ``` ``` ``` ``` ``` 是非常非常重要的 是非常非常重要的 是非常非常重要的更多关于更多关于更多关于更多关于 是非常
> 非常重要的 是非常非常重要的 是非常非常重要的 是非常非常重要的 是非常非常重要的 是非常非常重要的更多关于更多关于更多关于更多关于
> 更多关于更多关于更多关于更多关于更多关于更多关于更多关于更多关于更多关于更多关于更多关于更多关于更多关于更多关于更多关于更多关
> 于更多关于更多关于更多关于更多关于更多关于更多关于更多关于更多关于更多关于更多关于更多关于更多关于
> 
> <|user|>:
> 
>   <|assistant|>器和器和器和器和 ``` ``` ``` ```更多关于更多关于更多关于更多关于更多关于更多关于更多关于更多关于更多关于器和器和
> 器和器和器和器和 是非常非常重要的器和器和器和

其实上面文本中从"您感谢，只需要拨打拨打酒店的预订010-58576688预订或者......."这里往后的内容就都是无效的了，但是这并没有什么分隔的规律

谢谢

**Comment by toddlt** [comment]
而且evaluate过程中显存占用很异常，eval过程中显存占用会暴增到40GB到60GB且无法释放

**Comment by zRzRzRzRzRzRzR** [comment]
你这个输出确实不对啊，特殊token都不正常，至于你说的第二个问题，暂时没有遇到
微调的相关问题请到 #253 中咨询更多，我们定期在这个讨论中查看

**Comment by toddlt** [comment]
这个不是原始输出，我格式化打印的

**Comment by toddlt** [comment]
好的，我把问题记录在这里了 https://github.com/THUDM/ChatGLM3/discussions/253#discussioncomment-7650355

第二个问题，我看您这边的代码里没有写evaluate诶，trainer中只有train_dataset
```python
trainer = PrefixTrainer(
        model=model,
        args=training_args,
        train_dataset=train_dataset,
        tokenizer=tokenizer,
        data_collator=data_collator,
        save_changed=model_args.pre_seq_len is not None
)
```

**Comment by zRzRzRzRzRzRzR** [comment]
是的，确实没有写到验证，之前开梯度检查好像会有bug，所以我们只开了trainer，我们之后看以下，要补充

**Comment by zRzRzRzRzRzRzR** [comment]
是的，目前只有训练部分，具体的评估乐内容需要你自己根据你的指标来评估，因为没有固定的标签
